### PR TITLE
HDDS-7141. Recon: Improve Disk Usage Page

### PR DESCRIPTION
**Pull request by smitajoshi12**
## What changes were proposed in this pull request?

1) Make the DU page's texts the same font and color as on the other pages in Recon.
2) The chart description (which color stands for which volume/bucket/key) could be a bit bigger
3) If we click into the chart parts we go into that bucket and change the pie chart accordingly. If we click on the last key there is no pie chart showing up, we could change this to show a pie chart with 100% of the key.
4) There is a button on the right side that says Show Metadata for Current Path. If we click on it there is a page popping up on the right side with information about the current path. We could fix this page on the right side as we have plenty of space to do that. We can remove the button in this case and we don't need to make the left side of the page (the chart) grey.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-7141

## How was this patch tested?
Manually

### CODE REVIEW

**Comment by adoroszlai** [comment]
Whoever merges this, please make sure to edit the commit message to match PR title if necessary (because GitHub uses the message from commit by default if the PR has a single one).

**Comment by adoroszlai** [comment]
@dombizita @smengcl please review

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 1]
```diff
@@ -207,7 +220,7 @@ export class DiskUsage extends React.Component<Record<string, object>, IDUState>
     this.updatePieChart('/', DEFAULT_DISPLAY_LIMIT);
   }
 
-  clickPieSection(e, curPath: string): void {
+  clickPieSection(e, curPath: string): void {
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:223]: nit, please remove the unnecessary space here

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 2]
```diff
@@ -409,30 +422,44 @@ export class DiskUsage extends React.Component<Record<string, object>, IDUState>
                 {(duResponse.size > 0) ?
                   ((duResponse.size > 0 && duResponse.subPathCount === 0) ?
                     <div style={{height: 800}}>
+                    <div  className='metadatainformation'>
                       <br/> {' '}
-                      <h3>This object is a key with a file size of {this.byteToSize(duResponse.size, 1)}.<br/> {' '}
                         You can also view its metadata details by clicking the top right button.
-                      </h3>
-                    </div> :
+                    </div>
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:428]: I think we could remove this as well, as the plan with this patch was to default fix the metadata details on the right side and with the button you could close it.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 3]
```diff
@@ -409,30 +422,44 @@ export class DiskUsage extends React.Component<Record<string, object>, IDUState>
...
+                        font: {
+                          family: 'Roboto, sans-serif',
+                          size: 15
+                        },
+                        showlegend: true,
+                        title: 'Disk Usage for ' + returnPath + ' (Total Size: ' + this.byteToSize(duResponse.size, 1) + ')'
+                      }
+                    }/>
+                  </div>
+                    :
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:444]: I tried it and we could remove the second condition inside this nested condition. The only difference between these two `Plot`'s is the `You can also view its metadata details by clicking the top right button.` div, which can be removed and the `onClick`. We can add the condition inside the `onClick`, like this: `onClick={(duResponse.subPathCount === 0) ? undefined : e => this.clickPieSection(e, returnPath)}`
I also removed the `duResponse.size > 0` from the condition, as that is checked previously.

**Review by dombizita — CHANGES_REQUESTED** [review]
Thanks for the patch @smitajoshi12, I added some inline comments. Please fix the metadata details on right side default, but keep the button so we can hide it if we want to.

**Comment by kaijchen** [comment]
> Whoever merges this, please make sure to edit the commit message to match PR title if necessary (because GitHub uses the message from commit by default if the PR has a single one).

This could be changed in repo settings under "General - Pull Requests". However, it has to be done by the infra team.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 4]
```diff
@@ -152,26 +152,39 @@ export class DiskUsage extends React.Component<Record<string, object>, IDUState>
...
-        // The return subPath must be normalized in a format with
-        // a leading slash and without trailing slash
-        const pieces = subpath.path.split('/');
-        const subpathName = pieces[pieces.length - 1];
-        // Differentiate key without trailing slash
-        return (subpath.isKey || subpathName === OTHER_PATH_NAME) ? subpathName : subpathName + '/';
-      });
+      let pathLabels, values, percentage, sizeStr, pieces, subpathName;
+
+      if (duResponse.subPathCount === 0 || subpaths === 0) {
```
[the remark this answers is not in the document]
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:157]: HDDS-5377. Add a Namespace Summary tab to Recon Web UI . It is added by Yuan Ga to sort all  subpaths from End Point Response.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 5]
```diff
@@ -152,26 +152,39 @@ export class DiskUsage extends React.Component<Record<string, object>, IDUState>
...
+        pieces = duResponse && duResponse.path.split('/');
+        subpathName = pieces[pieces.length - 1];
+        pathLabels = [((duResponse.path).isKey || subpathName === OTHER_PATH_NAME) ? subpathName : subpathName];
```
[the remark this answers is not in the document]
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:160]: image

[the remark this answers is not in the document]
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:160]: Both are different variables  we are checking array length and subPathCount variable.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 6]
```diff
@@ -152,26 +152,39 @@ export class DiskUsage extends React.Component<Record<string, object>, IDUState>
...
+          pieces = subpath.path.split('/');
+          subpathName = pieces[pieces.length - 1];
+          // Differentiate key without trailing slash
+          return (subpath.isKey || subpathName === OTHER_PATH_NAME) ? subpathName : subpathName + '/';
+        });
 
-      const values = subpaths.map(subpath => {
-        return subpath.size / dataSize;
-      });
+        values = subpaths.map(subpath => {
```
[the remark this answers is not in the document]
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:175]: image

[the remark this answers is not in the document]
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:175]: Some Issue with Code Editor but looks same in editor.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 1 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:223]: Changed in Next Commit.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 2 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:428]: Could not Understand which Line  text need to remove siyao suggested to keep text.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 3 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:444]: There are two condition If  (duResponse.size > 0 && duResponse.subPathCount === 0)  as per siyao need to show empty Pie chart  no further clicks on this pie chart. If condition is false show regular pie chart as suggested by siyao.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (dombizita) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:157]: Oh okay, I misread that you are checking `duResponse.subPathCount`, not `duResponse.subPaths`. It's okay like this, thanks!

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (dombizita) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:160]: I was referring the condition in line 160, which defines pathLabels.
```suggestion
        pathLabels = subpathName;
```
This suggested change would work correctly, right?

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (dombizita) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:175]: I still see the plus spaces, so please check it again.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 2 above)
Comment (dombizita) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:428]: I can't suggest changes here, I thought to remove the whole div that starts on line 425.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 3 above)
Comment (dombizita) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:444]: The `duResponse.size > 0` check in the second condition is unnecessary, as we check it in the line above it in the first condition. That's why we can keep only `duResponse.subPathCount === 0` in the second condition. We can remove the Plot you added and add a change in the already existing Plot like this: `{(duResponse.subPathCount === 0) ? undefined : e => this.clickPieSection(e, returnPath)}` (and with this change we remove the "You can also view its metadata details by clicking the top right button." from the Plot as well, if we already planned to fix that)

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/index.less [hunk 7]
```diff
@@ -19,14 +19,11 @@
...
-    'Helvetica Neue', sans-serif;
+  font-family: 'Roboto', sans-serif;
   -webkit-font-smoothing: antialiased;
   -moz-osx-font-smoothing: grayscale;
 }
 
 code {
-  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
-    monospace;
+  font-family: 'Roboto', sans-serif;
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/index.less:28]: `code` is supposed to use monospace fonts correct? Why change it to Roboto?

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/index.less [hunk 8]
```diff
@@ -19,14 +19,11 @@
 body {
   margin: 0;
   padding: 0;
-  font-family: -apple-system, BlinkMacSystemFont, 'Roboto', 'Segoe UI',
-    'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans',
-    'Helvetica Neue', sans-serif;
+  font-family: 'Roboto', sans-serif;
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/index.less:22]: We should probably leave this untouched. And override this in the class's CSS instead if needed.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.less [hunk 9]
```diff
@@ -38,15 +38,46 @@
...
+    fill: rgba(0, 0, 0, 0.60) !important;
+    font-size: 17px !important;
+    line-height: 1.5;
+    font-feature-settings: 'tnum', "tnum";
+    font-variant: tabular-nums;
   }
 }
 
-.input {
+  .input {
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.less:66]: nit: indentation. along with others like this
```suggestion
.input {
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 10]
```diff
@@ -152,26 +152,39 @@ export class DiskUsage extends React.Component<Record<string, object>, IDUState>
...
-        // Differentiate key without trailing slash
-        return (subpath.isKey || subpathName === OTHER_PATH_NAME) ? subpathName : subpathName + '/';
-      });
+      let pathLabels, values, percentage, sizeStr, pieces, subpathName;
+
+      if (duResponse.subPathCount === 0 || subpaths === 0) {
+        pieces = duResponse && duResponse.path.split('/');
+        subpathName = pieces[pieces.length - 1];
+        pathLabels = [subpathName];
+        values = [0.1];
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:161]: Why the value 0.1?

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 11]
```diff
@@ -152,26 +152,39 @@ export class DiskUsage extends React.Component<Record<string, object>, IDUState>
...
         return (value * 100).toFixed(2);
       });
 
-      const sizeStr = subpaths.map(subpath => {
-        return this.byteToSize(subpath.size, 1);
-      });
+        sizeStr = subpaths.map(subpath => {
+          return this.byteToSize(subpath.size, 1);
+        }); 
+      }
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:186]: nit: indentation

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 12]
```diff
@@ -406,33 +419,32 @@ export class DiskUsage extends React.Component<Record<string, object>, IDUState>
                 </Col>
               </Row>
               <Row>
-                {(duResponse.size > 0) ?
-                  ((duResponse.size > 0 && duResponse.subPathCount === 0) ?
+                {(duResponse.size > 0)?
                     <div style={{height: 800}}>
+                    <div  className='metadatainformation'>
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:424]:
```suggestion
                    <div className='metadatainformation'>
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 2 above)
Comment (smengcl) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:427]: Ditto Zita. We can safely remove the second line as well.

**Review by smengcl — CHANGES_REQUESTED** [review]
Thanks @smitajoshi12 for the patch.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.less [hunk 13]
```diff
@@ -38,15 +38,46 @@
...
+    fill: rgba(0, 0, 0, 0.60) !important;
+    font-size: 17px !important;
+    line-height: 1.5;
+    font-feature-settings: 'tnum', "tnum";
+    font-variant: tabular-nums;
   }
 }
 
-.input {
+ .input {
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.less:66]: nit: indentation
```suggestion
.input {
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.less [hunk 14]
```diff
@@ -38,15 +38,46 @@
...
     padding-left: 5px;
     margin-right: 10px;
     display: inline-block;
     width: 300px;
-}
\ No newline at end of file
+  }
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.less:71]:
```suggestion
}
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.less [hunk 15]
```diff
@@ -38,15 +38,46 @@
...
+    font-size: 14px;
+    fill: rgba(0, 0, 0, 0.65) !important;
+    line-height: 1.5;
+    font-feature-settings: 'tnum', "tnum";
+  }
+  
+ .ant-drawer-open .ant-drawer-mask
+  {
+    opacity: 0 !important;
+  }
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.less:83]:
```suggestion
.metadatainformation {
  font-size: 14px;
  fill: rgba(0, 0, 0, 0.65) !important;
  line-height: 1.5;
  font-feature-settings: 'tnum', "tnum";
}

.ant-drawer-open .ant-drawer-mask
{
  opacity: 0 !important;
}
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 10 above)
Comment (smengcl) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:161]: Note: It is just a dummy value

**Comment by smengcl** [comment]
It looks like one of the `split` function is causing issues when the cluster is empty. @smitajoshi12 Would you check if this is the case on your docker cluster (when empty) as well?

Screen Shot 2022-10-18 at 10 10 25 AM

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 2 above)
Comment (smengcl) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx]: Done in ae1ff1d42a30e5e15a52946f99eed804f9615f5e

**Review by smengcl — APPROVED** [review]
Thank you @smitajoshi12 . I can confirm the latest commit fixes the UI error when the cluster is empty. I will merge this shortly.

**Pull request merged**